pkg/core/codeset_client.go

```diff
@@ -0,0 +1,199 @@
+package fuseml
+
+import (
+	"fmt"
+	"log"
+	"net/url"
+	"os/exec"
+	"path"
+	"time"
+
+	"code.gitea.io/sdk/gitea"
+	"github.com/pkg/errors"
+
+	codeset "github.com/fuseml/fuseml-core/gen/codeset"
+	giteac "github.com/fuseml/fuseml-core/pkg/core/gitea"
+)
+
+var (
+	// FIXME: generate this and put it in a secret
+	HookSecret = "generatedsecret"
+
+	// StagingEventListenerURL should not exist
+	// FIXME: detect this based on namespaces and services
+	StagingEventListenerURL = "http://el-mlflow-listener.fuseml-workloads:8080"
+
+	DefaultOrg = "workspace"
+)
+
+// CodesetClient provides functionality for talking to a
+// Fuseml installation on Kubernetes
+type CodesetClient struct {
+	giteaClient   *gitea.Client
```

Comment:
I wonder if the client indeed has to interact with the gitea API, or whether this is something that belongs on the server side, maybe as part of the codeset store implementation.

I would think that the CLI should only run those `git init ... git push` steps, because that's what a "regular" git user does anyway, while the rest (creating orgs, setting up webhooks and so on) should be done by the server. I mean, theoretically, these are two different things done through two separate APIs and even requiring different credentials/permissions: setting up the org/repository and pushing code to an existing repository.

What do you think about this separation of concerns: fuseml-core is a gitea admin user: it can access the gitea API and create/update/list projects (orgs) and create/update/list repositories (codesets), while the fuseml CLI acts as just a regular git client: it can only execute git operations (git clone, git push etc), but for everything else that requires access to the gitea API, it should go through the fuseml-core server ?

This separation plays well with another aspect: fuseml projects and codesets are FuseML specific "control plane" concepts, and the fuseml-core service is where they are centrally managed, so the CLI should call the server API for all operations that concern them. The only exception is the manipulation of codeset contents, which is just generic "payload" that can be transferred using another circuit - we can even call this the "data plane" - by using the git protocol that doesn't involve the fuseml-core API

It may just be me being heavily biased by my networking background, but this just feels like a better approach, what do you think ?

Reply:
Another argument I can bring in favor of this separation is that the client should be decoupled from the details of how the git server is implemented. It's gitea now, sure, but adding support for other git servers, or integrating FuseML with external git servers would be a lot cleaner if the implementation was managed server-side, while this is transparent for the client, given that it only "speaks" the generic git protocol expected to be implemented by any git server .

Reply:
I agree that generally, whenever client is not required, it would be better to do the git operations from server.

So yes, as you mention: org manipulation and webhook creation could be done on server (as well as pulling for the sake of pipelines if they'd need it), while actual pushing the code needs to happen from client.

One thing I'm not sure about: how to pass the credentials? Do you mean server side should use the admin ones (like they are being used right now in this code) and client side should have specific user only ...
On little problem is that first such user would need to be created, so the `codeset register` would need to be split into 2 steps. But OK.

Next question is, how should I pass the credentials for the server side? Should it be part of the query? But that it means used would need to know it anyway. fuseml/carrier does it by checking the `secret` from gitea deployment, i.e. directly accessing kubernetes cluster. While I wanted to avoid this on client side, it could be fine on server...

As for gitea: so far I like the gitea API. As you can see, the client part are currently done with just executing git client though (but I'd like to change it to gitea too). But indeed, it might be good for the future to get independent.

Reply:
I think the key to differentiating between what the client does directly and what the client does through the server is this: there are in fact two APIs exposed by the gitea server at work here, one is the gitea API that works _only_ with gitea (described in detail here: https://try.gitea.io/api/swagger#/), the other is the git protocol, that works with _any_ git server. Conversely, there are two types of client: the gitea client and the git client.

To keep the fuseml client implementation independent of the flavor of git server being used, only the git protocol should be used by the client when talking to the git server directly. This is sufficient for git repository operations like `git push` and `git pull` that are needed to transfer the codeset contents to/from the git server. All other operations that require access to the gitea API - things like managing (creating/deleting/configuring) organizations, repositories, credentials, webhooks etc. - should not be done by the client, but by the server instead. As mentioned in [my wiki PR](https://github.com/fuseml/wiki/pull/1), the fuseml client calls the FuseML server API to create a project (represented by a git organization) or a codeset (represented by a git repository), and it is the FuseML server which calls the gitea API to create those things.

> One thing I'm not sure about: how to pass the credentials? Do you mean server side should use the admin ones (like they are being used right now in this code) and client side should have specific user only ...
> On little problem is that first such user would need to be created, so the codeset register would need to be split into 2 steps. But OK.

Yes, when the server talks to the gitea API, it uses gitea's admin account (or another account with admin privileges). The client needs another set of credentials, one that is only required to authorize the client for `git push` operations in that repository (`git fetch/pull` being un-authenticated for "public" repositories anyway, same as github). To avoid having to implement the whole auth/authz feature just for this, I guess you could just create a hard-coded gitea user - let's call this the `gitops` user - and give it push access to all repositories. Either that, or configure codeset repositories so that you don't need to authorize users for push operations, if that's possible. You could even transfer the `gitops` user credentials back to the client in the POST response, if that makes things clearer from a client perspective.

> Next question is, how should I pass the credentials for the server side? Should it be part of the query? But that it means used would need to know it anyway. fuseml/carrier does it by checking the secret from gitea deployment, i.e. directly accessing kubernetes cluster. While I wanted to avoid this on client side, it could be fine on server...

I'm not sure I understand this question, but maybe my previous comments already help answer it. We can keep a single hard-coded user account to be used for clients when pushing code to any repository.

> As for gitea: so far I like the gitea API. As you can see, the client part are currently done with just executing git client though (but I'd like to change it to gitea too). But indeed, it might be good for the future to get independent.

I don't think the gitea API was ever meant to replace the git client commands. Yes, you can PUT files in a gitea repository using the gitea API, but it's not meant to be used instead of `git push`.

Reply:
> I'm not sure I understand this question, but maybe my previous comments already help answer it. We can keep a > single hard-coded user account to be used for clients when pushing code to any repository.

I'm not talking about client side, but server side, which needs that gitea specific access. Currently I'm getting credentials from env variables on client side.  Should I 
- use different set of env variables on server side (for the admin user etc.)?
- pass the data through the REST API?
- fetch the gitea information from Kuberneres cluster

> I think the key to differentiating between what the client does ...

Yes, I agree that the client part does not need to be gitea specific as I already mentioned. But I'd keep it for next iteration.

> You could even transfer the gitops user credentials back to the client in the POST response, if that makes
> things clearer from a client perspective.

That was my original idea indeed. But maybe we can do with static user instead for now, as you suggest.

Reply:
> I'm not talking about client side, but server side, which needs that gitea specific access. Currently I'm getting credentials from env variables on client side.

The way I've seen this usually solved in cloud native applications is by having the fuseml-core service read the gitea admin user credentials from a configuration file, or better, from environment variables. How those env variables are passed onto the fuseml-core container depends on the method of installation. For example, if we had an uber-helm chart overseeing the configuration of all services that make up the fuseml app, I imagine the helm chart would have a global option holding the admin gitea password, which it would then use to configure both gitea itself and the fuseml-core service. Or it could generate the password, same as we did for SPR.

Reply:
OK, I can settle on env variables that server reads for now. It could evolve later.